plugin/template

rtreffer opened the pull request:
The template plugin matches the incoming query by class, type and regex and templates a response with a go templates.

As suggested in #1289

A README.md is in place. I haven't touched the root README.md.

Missing: 
1. ~~@miekg I haven't changed the rcode directive yet. Can you think of any rcode besides NXDOMAIN and ServFail that might be useful? In that case I'd go with `error|denial`.~~
2. ~~Metrics (esp. internal failures to template)~~

miekg commented:
Besides NXDOMAIN there is also NODATA (name exist, type does not). But NODATA does not have an rcode it is just SUCCESS; so you need that anyway which means you'll need `denial`.
SERVFAIL, REFUSED, NOTIMPL are a few server failures that come up mind.

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,135 @@
+# template
+
+*template* allows for dynamic responses base on the incoming query.
```
*based

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,135 @@
+
+~~~
+template CLASS TYPE [REGEX...] {
+    [answer section]
```
is `answer` a keyword? should this be `answer SECTION`?
(not sure how to read this yet)

miekg replied:
no authoritity section?

miekg replied:
or better `answer RR`

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,135 @@
+* **CLASS** the query class (usually IN or ANY)
+* **TYPE** the query type (A, PTR, ...)
+* **REGEX** regex that are matched against the incoming query. Specifying no regex matches everything.
+* `section` A RFC 1035 style fragment build by a go template that contains the answer.
```
*Go

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,135 @@
+## Syntax
+
+~~~
+template CLASS TYPE [REGEX...] {
```
what happens when multiple template match? First one wins?

rtreffer replied:
Added to the README

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,135 @@
+
+### Resolve .invalid as NXDOMAIN
+
+The `.invalid` domain is a reserved TLD to indicate non-existing domains.
```
really? Didn't know that (can add a RFC reference, i.e. See RFC XXXX)

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,135 @@
+
+Imagine you run `example.com` with a datacenter `dc1.example.com`. The datacenter domain
+is part of the DNS search domain.
+However `something.example.com.dc1.example.com` would usually indicate a wrong autocomplete.
```
autocomplete -> indicates a non fully qualified domain that inadvertely has the default domain added or search path added.

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,135 @@
+}
+~~~
+
+An IP consists of 4 bytes, `a.b.c.d`. Named groups make it less error prone to reverse the
```
An v4 IP address

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,135 @@
+    # ip-a-b-c-d.example.com A a.b.c.d
+
+    template IN A (^|[.])ip-10-(?P<b>[0-9]*)-(?P<c>[0-9]*)-(?P<d>[0-9]*)[.]example[.]$ {
+      answer "{{ .Name }} 60 IN A 10.{{ .Group.b }}.{{ .Group.c }}.{{ .Group.d }}"
```
No RCODE mean SUCCESS?

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,135 @@
+      additional "{{ .Name }} 60 IN A 10.{{ .Group.b }}.{{ .Group.c }}.{{ .Group.d }}"
+    }
+}
+~~~
```
Might be worth adding a "See Also" section that points to Go templates, godoc.org link? And maybe RFC 1035 for the text presentation?

Also add BUG section that duplicates the warning in the intro about caddy file eating the templates when the spacing is off.

miekg commented on plugin/template/setup.go:
```diff
@@ -0,0 +1,127 @@
+
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+	"github.com/mholt/caddy"
```
newline between "our" imports and third party.

miekg commented on plugin/template/setup.go:
```diff
@@ -0,0 +1,127 @@
+			return nil, c.ArgErr()
+		}
+
+		class, found := dns.StringToClass[c.Val()]
```
typical go is to use `ok` instead of found.

miekg commented on plugin/template/setup.go:
```diff
@@ -0,0 +1,127 @@
+
+		class, found := dns.StringToClass[c.Val()]
+		if !found {
+			return nil, c.Errf("Invalid query class %s", c.Val())
```
Errors always start with a lowercase

miekg commented on plugin/template/setup.go:
```diff
@@ -0,0 +1,127 @@
+		}
+		queryType, found := dns.StringToType[c.Val()]
+		if !found {
+			return nil, c.Errf("Invalid query rr type %s", c.Val())
```
drop 'query' and uppercase rr -> RR

miekg commented on plugin/template/template.go:
```diff
@@ -0,0 +1,138 @@
+	"context"
+	"regexp"
+	"strconv"
+
```
drop this newline

miekg commented on plugin/template/template.go:
```diff
@@ -0,0 +1,138 @@
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
```
newline above miekg/dns

miekg commented on plugin/template/template.go:
```diff
@@ -0,0 +1,138 @@
+	return "template"
+}
+
+func executeRRTemplate(template *gotmpl.Template, data templateData) (*dns.RR, error) {
```
why does this return a pointer to dns.RR?

miekg commented on plugin/template/template.go:
```diff
@@ -0,0 +1,138 @@
+		return nil, err
+	}
+	rrDefinition := buffer.String()
+	rr, err := dns.NewRR(rrDefinition)
```
`rr, err := dns.NewRR(buffer.String())`

miekg commented on plugin/template/template.go:
```diff
@@ -0,0 +1,138 @@
+}
+
+func (t template) match(r *dns.Msg) (templateData, bool) {
+	for _, q := range r.Question {
```
There is only 1 question in the question section (long story), so there is no need to range. CoreDNS quarantees you have one before calling into the plugins.

rtreffer commented:
@miekg I've added some metrics, I think I've addressed all your concerns. I'm dropping the WIP, can you do another review? (I was less available over the holidays than I've hoped)

johnbelamaric commented on plugin/template/README.md:
```diff
@@ -0,0 +1,162 @@
+* `.Class` the query class (usually `IN`)
+* `.Type` the RR type requested (e.g. `PTR`)
+* `.Match` an array of all matches. `index .Match 0` refers to the whole match.
+* `.Group` a map fo the named capture groups.
```
s/fo/of/

rtreffer replied:
fixed

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,166 @@
+
+* **CLASS** the query class (usually IN or ANY)
+* **TYPE** the query type (A, PTR, ...)
+* **REGEX** [Go regexp](https://golang.org/pkg/regexp/) that are matched against the incoming query. Specifying no regex matches everything (default: `.*`). First matching regex wins.
```
incoming "question name"?

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,166 @@
+The output of the template must be a [RFC 1035](https://tools.ietf.org/html/rfc1035) style resource record line (commonly refered to as a "zone file").
+
+**WARNING** there is a syntactical problem with Go templates and caddy config files. Expressions like `{{$var}}` will be interpreted as a reference to an environment variable by caddy/coredns while `{{ $var }}` will work. Try to avoid template variables. See [Bugs](#bugs)
+
```
Final dot is missing.

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,166 @@
+## Metrics
+
+If monitoring is enabled (via the *prometheus* directive) then the following metrics are exported:
+- `coredns_template_matches_total` the total number of matched requests.
```
please include in labels in `{}` in the metric name. see other READMEs, if there are non just use `{}`

miekg replied:
please do this before we can merge.

rtreffer replied:
fixed

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,166 @@
+- `coredns_template_template_failures_total` the number of times the Go templating failed.
+- `coredns_template_rr_failures_total` the number of times the templated resource record was invalid and could not be parsed.
+
+Both failure cases indicate a problem with the template configuration.
```
Are these runtime errors? I.e. can we catch the template config errors on startup?

rtreffer replied:
Yes, even if the template is syntactical correct it's hard to tell if
1. it works with the input
2. the output is a valid RR

Regex and template syntax errors will break startup, though.

miekg commented on plugin/template/template.go:
```diff
@@ -0,0 +1,143 @@
+
+// ServeDNS implements the plugin.Handler interface.
+func (h Handler) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	for _, template := range h.Templates {
```
This needs a check if you should handle the zone in the first place and if not call NextOrFailure here.

rtreffer replied:
Where would the plugin get a zone? It handles regex only (for now)

miekg replied:
```
	zone := plugin.Zones(e.Zones).Matches(state.Name())
	if zone == "" {
		return plugin.NextOrFailure(e.Name(), e.Next, ctx, w, r)
	}
```
as copied from the etcd plugin (there is more code there, take a look)

rtreffer replied:
Yes, I saw that, but there is no Zone for this plugin: it acts on regex rather than zone which is a more sophisticated match. There is no `e.Zones` equivalent right now.

If nothing matches `return plugin.NextOrFailure()` is done. If a regex matches the answer will be generated (or a servfail if that causes an error).

rtreffer replied:
So the question is mainly: should this plugin actually act on zones and use a zone scope? Or is "try to match everything (the regex can limit the zone anyway)"?

Or should this be changed to act on zones and move the regex into a directive?

miekg commented on plugin/template/template.go:
```diff
@@ -0,0 +1,143 @@
+			}
+			msg.Answer = append(msg.Answer, rr)
+		}
+		for _, additional := range template.additional {
```
Why are we excluding the authority section? I mean that's fine, but it should be documented?

rtreffer replied:
Well, I'm unsure what all side effects of that are.

1. Do we need flag setting in that case, too?
2. Should we allow anything in the authority section?

I would wait for a use case on that, I don't have one so I'm unsure about it.

miekg replied:
I don't know; it up to the user; if you write crap in there... garbage in garbage out? You can shoot yourself in the foot already. We can postpone adding this in a follow up PR.

rtreffer replied:
Nah, in that case it's easy, I'll add it.

rtreffer replied:
authority section pushed.

miekg commented on plugin/template/template.go:
```diff
@@ -0,0 +1,143 @@
+	buffer := &bytes.Buffer{}
+	err := template.Execute(buffer, data)
+	if err != nil {
+		TemplateFailureCount.WithLabelValues(data.Regex, section, template.Tree.Root.String()).Inc()
```
Slightly nice to let the caller increment the failure counter? I.e similar to not log errors in a function, but just returning the error?

rtreffer replied:
This function combines 2 steps and their error handling.

Unrolling the error in the caller is close to inlining this function in the first place. Happy to do that if it makes sense.

miekg replied:
nah, leave it for this then.

miekg commented on plugin/template/template.go:
```diff
@@ -0,0 +1,143 @@
+	return rr, nil
+}
+
+func (t template) match(r *dns.Msg) (templateData, bool) {
```
I think is nicer and more inline with other template to create a `state := request.Request(...)` and use that throughout. Give you some small helper functions that speeds up things, no need to keep lowercasing the question name over and over.

rtreffer replied:
Oh no, I was missing the ToLower piece completely. The input is now a a request.Request and state.Name() is used to retrieve the qname.

miekg commented on plugin/template/README.md:
```diff
@@ -0,0 +1,190 @@
+template CLASS TYPE [REGEX...] {
+    [answer RR]
+    [answer RR]
+    [additional RR]
```
this needs authortity as well now.

rtreffer replied:
fixed

rtreffer commented:
@miekg next round of fixes.

The only thing I'm still missing is the zone limiting. The plugin currently operates on `regex` as limiting factor and not any particular zone (right now `template <regex>` instead of `template <zone> { regex <regex>`). This is more powerful but not in line with other plugins. (Matching `something.example.$` does set a zone). Fallthrough is currently always on. This is useful to template e.g. ip templates under a shared domain.

1. I could explain/highlight the current behavior in the documentation
1. I could externalize the main body after a match, making the regex matching / fallback logic obvious to read
1. I could change the plugin to be bound to zone(s) and add a configurable fallthrough behavior (this is the same as a match-all NXDOMAIN regex)

^ @miekg thoughts? 3.?

johnbelamaric commented:
You can limit it by zone, and make the regex a directive. That way it does a fast filter out of the irrelevant queries. You can always make the zone "." if you want it to check every query.

miekg commented:
+1 to what @johnbelamaric said.

Also we've decided to make `fallthrough` always explicit (in the other plugins)

miekg commented:
> I could change the plugin to be bound to zone(s) and add a configurable fallthrough behavior (this is the same as a match-all NXDOMAIN regex)
I think this is a sensible thing to do.

Bit tricky to fiddle that in the syntax as it should be optional (then default to the server's block zones):
`template [ZONES] CLASS TYPE [REGEX...] {`

Should work by checking for IN/CH/ANY as a class, which you then can't have as zonenames, by meh,

gibson042 commented on plugin/template/setup.go:
```diff
@@ -0,0 +1,143 @@
+			return nil, c.ArgErr()
+		}
+
+		class, ok := dns.StringToClass[c.Val()]
```
As noted in https://github.com/miekg/dns/issues/577#issuecomment-346235028 , "ANY" is not a real class mnemonic. This should be prepared for its removal, e.g.
```go
if c.Val() == "ANY" {
	t.class = dns.ClassANY
} else {
	class, ok := dns.StringToClass[c.Val()]
	if !ok {
		return nil, c.Errf("invalid query class %s", c.Val())
	}
	t.class = class
}
```

rtreffer replied:
The linked page lists ANY: `255 QCLASS * (ANY)`
https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-2

ANY is not ALL, so you are free to respond with nothing.

rtreffer replied:
Ah, removing from the list, yeah, that's fine, we just have to handle the ordering of these 2 PRs

gibson042 replied:
btw, ANY is also an invalid RTYPE mnemonic so you'll want similar code for that.

rtreffer replied:
It is valid for querying and matching. It's '*' or 'ANY'.
It is not valid in DNS responses, **but these values here are never used in RRs, they are used for matching**.

gibson042 replied:
I feel like you're misunderstanding me... ANY is not valid for either class or type in an RFC 1035 master file, even though the values are assigned and CLASS255 or TYPE255 _would_ be valid (if nonsensical). You're relying on an entry in the `StringToClass` map that doesn't belong there, and you should stop doing so in order to avoid breaking if the bug in the base library is fixed.

rtreffer replied:
Yes, I totally understand that, but they are not used in the context of 1035 here. They are not even used in the context of an RR here.

They are used to *match incoming types*. The same way you can ask a server to return *something* by sending an ANY query.

gibson042 replied:
I'm not telling you to drop "ANY" from your matchers, I'm telling you to stop assuming that "ANY" will be in `StringToClass` (which _is_ used for parsing presentation format).

rtreffer replied:
Which is your suggestion in another PR, I know that.

But I could also invert that: Why not fix the parsing to only use mnemonic of that context?

The documents you pull as reference call it ANY. It is commonly referred to as ANY. It is just not valid in a certain, different, context.

Anyway, fixing this either way is trivial. I don't see any value in the debate here.

rtreffer commented:
@miekg I would in that case rather move the regex into a directive. Perhaps as "match"....

```
template example. {
  match IN A "....."
  anwser .....
}
```

Does that look reasonable?

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/templa..." ]
>gibson042 commented on this pull request.
>
>
>
>> +		return Handler{Next: next, Templates: templates}
>+	})
>+
>+	return nil
>+}
>+
>+func templateParse(c *caddy.Controller) (templates []template, err error) {
>+	templates = make([]template, 0)
>+
>+	for c.Next() {
>+		t := template{}
>+		if !c.NextArg() {
>+			return nil, c.ArgErr()
>+		}
>+
>+		class, ok := dns.StringToClass[c.Val()]
>
>As noted in https://github.com/miekg/dns/issues/577#issuecomment-346235028 , "ANY" is not a real class mnemonic. This should be prepared for its removal, e.g.
>```go
>if c.Val() == "ANY" {
>	t.class = dns.ClassANY
>} else {
>	class, ok := dns.StringToClass[c.Val()]
>	if !ok {
>		return nil, c.Errf("invalid query class %s", c.Val())
>	}
>	t.class = class
>}
>```

I'm not sure we're going to do that though, so it's fine to leave this as is.

gibson042 commented:
> I'm not sure we're going to do that though, so it's fine to leave this as is.

It relies upon a bug in the base library that is trivial to avoid... why would you want to keep code that will break if the bug is fixed?

miekg commented:
I propose merging this and iterating in a couple of new PRs.
(I also want to release 1.0.3 and have this plugin hidden until we call it final, this means not putting it in plugin.cfg for a short while)

rtreffer commented:
@miekg what's your planned timeline for 1.0.3? Today? Tomorrow?

miekg commented:
either of those :)

Also I can't read the PR anymore because of the stupid way github does comments

rtreffer commented:
So the only objection I would have is that this means the syntax would change afterwards again (to add zones)

miekg commented:
mostly I would guess. I'll also take a closer look and either file (smallish) bugs and/or PRs
